### PR TITLE
Feature/3661 active subscription check

### LINKED ISSUE #3661: Delete site - disallow sites with active subscription from deleting their site
Sites with an active subscription must remove their subscriptions before they can delete their site.

We can link them to their subscriptions in a webview that they can handle.

### PR DESCRIPTION
**Pull request by kwonye**
Fixes #3661

To test:
1. Have a site that has active purchases on it and try to delete site
2. There will be a pop up that says you can't delete a site with active purchases
3. Click "Show purchases" and it will open up a webview
4. Delete your active purchases
5. Go back and delete your site again and the confirmation dialog will show up and you can delete the site just fine.

Needs review: @oguzkocer and @nbradbury. Thanks guys!

### CODE REVIEW

**Comment by kwonye** [comment]
Also there is some work in redesigning the feature (follow conversation here: #3648). All UI/text requests should go there please.

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java [hunk 1]
```diff
@@ -384,9 +382,8 @@ public boolean onPreferenceClick(Preference preference) {
         } else if (preference == mCategoryPref || preference == mFormatPref) {
             return !shouldShowListPreference((DetailListPreference) preference);
         } else if (preference == mExportSitePref) {
-            showExportContentDialog();
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java]: I am sorry if I missed a conversation about this. Why are we removing this? Also, if we are removing it, I think we should also remove the if statement: `else if (preference == mExportSitePref) {` since it's empty now.

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java [hunk 2]
```diff
@@ -384,9 +382,8 @@ public boolean onPreferenceClick(Preference preference) {
...
         } else if (preference == mDeleteSitePref) {
-            showDeleteSiteDialog();
+            requestPurchasesForDeletionCheck();
         } else {
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java]: I know this is not in the changes, but this `else` here is not necessary, I think we should just `return false` in the end. Since it's a very small change, can we add it to this PR?

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java [hunk 3]
```diff
@@ -384,9 +382,8 @@ public boolean onPreferenceClick(Preference preference) {
         } else if (preference == mCategoryPref || preference == mFormatPref) {
             return !shouldShowListPreference((DetailListPreference) preference);
         } else if (preference == mExportSitePref) {
-            showExportContentDialog();
         } else if (preference == mDeleteSitePref) {
-            showDeleteSiteDialog();
+            requestPurchasesForDeletionCheck();
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java]: I don't know if there is a separate issue/PR for this, but I think we should show some indication that we're making a request to the server. When the user taps on "Delete site", especially on a slow connection, it looks like nothing is happening.

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java [hunk 4]
```diff
@@ -692,20 +689,67 @@ public String format(int value) {
...
+        WordPress.getRestClientUtils().getSitePurchases(currentBlog.getDotComBlogId(), new RestRequest.Listener() {
+            @Override
+            public void onResponse(JSONObject response) {
+                showPurchasesOrDeleteSiteDialog(response, currentBlog);
+            }
+        }, new RestRequest.ErrorListener() {
+            @Override
+            public void onErrorResponse(VolleyError error) {
+                ToastUtils.showToast(getActivity(), getString(R.string.purchases_request_error));
+                AppLog.e(AppLog.T.API, error.toString());
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java]: Maybe add a description for the error here, something like `"Error occurred while requesting purchases for deletion check: " + error.toString()`. I think it'd be easier for us to find the issue from the logs, especially for the HEs.

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java [hunk 5]
```diff
@@ -692,20 +689,67 @@ public String format(int value) {
...
+    private void showPurchasesOrDeleteSiteDialog(JSONObject response, final Blog currentBlog) {
+        try {
+            JSONArray purchases = response.getJSONArray(PURCHASE_ORIGINAL_RESPONSE_KEY);
+            if (hasActivePurchases(purchases)) {
+                showPurchasesDialog(currentBlog);
+            } else {
+                showDeleteSiteDialog();
+            }
+        } catch (JSONException e) {
+            AppLog.e(AppLog.T.API, e.toString());
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java]: Similar to the error message comment above :) Sorry for the nitpicks!

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java [hunk 6]
```diff
@@ -692,20 +689,67 @@ public String format(int value) {
...
+            @Override
+            public void onErrorResponse(VolleyError error) {
+                ToastUtils.showToast(getActivity(), getString(R.string.purchases_request_error));
+                AppLog.e(AppLog.T.API, error.toString());
+            }
+        });
+    }
+
+    private void showPurchasesOrDeleteSiteDialog(JSONObject response, final Blog currentBlog) {
+        try {
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java]: I have a question for my own benefit here, I hope you don't mind. Is there any chance the `response` could be `null` here? Is it worth to check here?

Note: I am still trying to wrap my head around the whole NPEs and keep missing them in my own PRs, so just trying to understand it.

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java [hunk 7]
```diff
@@ -692,20 +689,67 @@ public String format(int value) {
...
             @Override
             public void onClick(DialogInterface dialog, int which) {
                 dialog.dismiss();
             }
         });
-
         builder.show();
     }
 
+    private boolean hasActivePurchases(JSONArray purchases) throws JSONException {
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java]: Similar to above, should we do a `null` check here? It looks like `getJSONArray` could return null from here: http://docs.oracle.com/javaee/7/api/javax/json/JsonObject.html#getJsonArray-java.lang.String-


P.S: I am ready to jump on the hate train for the NPEs. `Kotlin` anyone?

**Comment by oguzkocer** [comment]
Hey @kwonye! I left a few comments, I hope they are helpful. Please feel free to ignore any of them if they don't make sense, I tried my best to do a detailed review this time around.

I've tested the flow and the subscription check is working fine. However, you can not delete a site. I have debugged this a bit and it looks like the problem is caused by passing `null` as the data in `DeleteSiteDialogFragment` [here](https://github.com/wordpress-mobile/WordPress-Android/blob/feature/3661-active-subscription-check/WordPress/src/main/java/org/wordpress/android/ui/prefs/DeleteSiteDialogFragment.java#L85) and checking for `null` in the `SiteSettingsFragment` [here](https://github.com/wordpress-mobile/WordPress-Android/blob/feature/3661-active-subscription-check/WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java#L245), so `deleteSite()` is never called.

I am done for my first pass, let me know when you want me to take another look!

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java (hunk 1 above)
Comment (kwonye) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java]: Ah, that's my fault. I didn't want to add too many features in a single PR. I removed all export content work and will be doing it as a seperate PR. Good eye! :eyes:

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java (hunk 2 above)
Comment (kwonye) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java]: Definitely

Comment (kwonye) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java]: Actually I do think this is required. Some of the branches do not return, therefore the `return true` at the end.

We only absolutely return false if it didn't go into any of those conditionals.

Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java]: That's my bad, sorry about that!

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java (hunk 3 above)
Comment (kwonye) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java]: Good call, added.

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java (hunk 6 above)
Comment (kwonye) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java]: There is no chance it's null because `RestRequest` makes sure of it here: https://github.com/Automattic/android-wordpress-com-rest/blob/master/WordPressComRest/src/main/java/com/wordpress/rest/RestRequest.java#L112

If it can't parse the JSONObject (or it's null), it catches it and creates a new Response.Error object and returns that. So it's either good JSON or an Error.

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java (hunk 7 above)
Comment (kwonye) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java]: > P.S: I am ready to jump on the hate train for the NPEs. Kotlin anyone?

:laughing:

**Comment by kwonye** [comment]
> I've tested the flow and the subscription check is working fine. However, you can not delete a site. I have debugged this a bit and it looks like the problem is caused by passing null as the data in DeleteSiteDialogFragment here and checking for null in the SiteSettingsFragment here, so deleteSite() is never called.

Woah, that's not good... I'll figure out what's going on. Thanks for the great review!

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java (hunk 6 above)
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java]: That's great to know! Thanks for the explanation :)

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java (hunk 7 above)
Comment (maxme) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java]: > P.S: I am ready to jump on the hate train for the NPEs. Kotlin anyone?

Screenshot taken, I amass evidences.

**Comment by kwonye** [comment]
@oguzkocer null issue fixed, thanks!

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/DeleteSiteDialogFragment.java [hunk 8]
```diff
@@ -76,7 +83,7 @@ public void onClick(DialogInterface dialog, int which) {
             public void onClick(DialogInterface dialog, int which) {
                 Fragment target = getTargetFragment();
                 if (target != null) {
-                    target.onActivityResult(getTargetRequestCode(), Activity.RESULT_OK, null);
+                    target.onActivityResult(getTargetRequestCode(), Activity.RESULT_OK, new Intent());
```
Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/DeleteSiteDialogFragment.java]: Do you think it would be better if we kept the data `null` and instead check for this outside of `if (data != null)` statement? I don't think we care about the data, and someone might accidentally remove the `new Intent()` since it "looks" kind of pointless here. It's not a big deal but it also uses extra memory (a tiny bit, that is :P )

Comment (kwonye) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/DeleteSiteDialogFragment.java]: It was tricky, but I think you're right. I had a fear of readability in `onActivityResult`, but it looks fine even with the multiple branches. Addressed in c1e4f86dfa3ea4990cf020df21e2e11535e1fef4

**Comment by oguzkocer** [comment]
Looks good to me :shipit:

@nbradbury Let me know if you want to review it as well.

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java [hunk 9]
```diff
@@ -702,20 +702,70 @@ public String format(int value) {
         });
     }
 
-    private void showExportContentDialog() {
+    private void requestPurchasesForDeletionCheck() {
+        final Blog currentBlog = WordPress.getCurrentBlog();
+        final ProgressDialog progressDialog = ProgressDialog.show(getActivity(), "", getString(R.string.checking_purchases), true, false);
+        WordPress.getRestClientUtils().getSitePurchases(currentBlog.getDotComBlogId(), new RestRequest.Listener() {
+            @Override
+            public void onResponse(JSONObject response) {
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java]: It's possible for the network response/error to occur when the fragment is no longer attached, so it's important to check `isAdded()` in both `onResponse()` and `onErrorResponse()`.

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java (hunk 7 above)
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java]: > Screenshot taken, I amass evidences.

:astonished:

File: WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java (hunk 9 above)
Comment (kwonye) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java]: Due to the un-cancelable ProgressDialog, I didn't think that would be the case, but I trust your call on this one. Added in 71999d9db45fdb43a928cffba1b83a12029f00cc

Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java]: > Due to the un-cancelable ProgressDialog

Ah, sorry - hadn't noticed the dialog was un-cancelable.

Comment (oguzkocer) [inline WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java]: It might still be a good idea to check to make it future proof. `SiteSettingsFragment` is a beast of a file, we need all the help we can get :)

**Comment by nbradbury** [comment]
For testing purposes, I modified `hasActivePurchases` to always return `true` so I could see what happens when there are active purchases. After I tapped "Show Purchases" on the dialog, it took me to this WebView:

device-2016-02-19-184652

This shows all my purchases, which in my case are all on my primary domain (which is *not* the site I was deleting). Am I correct in assuming the end user is going to see a similar list, which isn't necessarily a list of purchases on the site they're deleting?

**Comment by kwonye** [comment]
> Am I correct in assuming the end user is going to see a similar list, which isn't necessarily a list of purchases on the site they're deleting?

The search only looks at purchases active on the current site, so they will never see that page from clicking the "delete site" preference. As shown here: https://github.com/wordpress-mobile/WordPress-Android/blob/feature/3661-active-subscription-check/libs/networking/WordPressNetworking/src/main/java/org/wordpress/android/networking/RestClientUtils.java#L270

**Comment by nbradbury** [comment]
> The search only looks at purchases active on the current site, so they will never see that page from clicking the "delete site" preference

But if the end user *does* has active purchases on the site they're deleting and they click "Show purchases," are they only going to see purchases on the blog they're wanting to delete? Based on [this call](https://github.com/wordpress-mobile/WordPress-Android/blob/feature/3661-active-subscription-check/WordPress/src/main/java/org/wordpress/android/ui/prefs/SiteSettingsFragment.java#L748) I think they're going to see active purchases on any of their sites.

**Comment by nbradbury** [comment]
Related to our Slack conversation about using WebViews in the app, is it possible to show the purchases page without the navigation icons? Users who tap one of those navigation icons will be taken to another page without a way to easily get back to the purchases page, and they shouldn't be expected to understand they're viewing our web app inside our native app.

76db36d4-d739-11e5-82a0-0e51a2eba47a

File: WordPress/src/main/res/values/strings.xml [hunk 10]
```diff
@@ -1301,6 +1301,11 @@
     <string name="delete_site_summary">This action can not be undone. Deleting your site will remove all content, contributors, and domains from the site.</string>
     <string name="delete_site_hint">Delete your site</string>
     <string name="pref_key_site_start_over_screen">wp_pref_key_site_start_over_screen</string>
-    <string name="delete_site_progress">Deleting site...</string>
+    <string name="delete_site_progress">Deleting site…</string>
+    <string name="purchases_request_error">Something went wrong. Please contact support</string>
```
Comment (nbradbury) [inline WordPress/src/main/res/values/strings.xml]: We shouldn't say "Please contact support" without providing an obvious, immediate way to do so. In this case I think the feature isn't going to be widely used, and the error isn't very likely, so a generic error message without the support mention should suffice.

**Comment by maxme** [comment]
> Related to our Slack conversation about using WebViews in the app, is it possible to show the purchases page without the navigation icons?

I think we should hide this navigation buttons in all Webviews of the app that shows a Calypso page.

**Comment by kwonye** [comment]
@nbradbury and @maxme the discussion for hiding the masterbar has been started here: https://github.com/Automattic/wp-calypso/issues/3353

I don't believe this should block the feature because the amount of users that see this page in the first place will be a low percentage (Most users do not have purchases on their site)

**Comment by nbradbury** [comment]
> I don't believe this should block the feature because the amount of users that see this page in the first place will be a low percentage

Agreed. FWIW, I also agree with Matias that a separate page for rendering just purchases makes more sense than a query param to hide the masterbar.

**Comment by kwonye** [comment]
> I meant to add that I do consider showing purchases for all sites rather than just the one being deleted to be a blocker.

@nbradbury I disagree. I'll look into how to back up this claim, but I believe users who have purchases on multiple sites will be able to understand the page showing all their purchases. Also, (again, I'll find the numbers) I believe only about 1% of users have purchases on more than one site.

**Comment by nbradbury** [comment]
@kwonye I guess the simple question is, "what does the web app do?" If the web app also shows you a list of purchases from all your sites, then I guess we go with it even though I don't like it :)

**Comment by kwonye** [comment]
> @kwonye I guess the simple question is, "what does the web app do?" If the web app also shows you a list of purchases from all your sites, then I guess we go with it even though I don't like it :)

The web does the same

**Comment by nbradbury** [comment]
> The web does the same

In that case, :shipit: :)
